tests/smoke/gmres.cpp

```diff
@@ -0,0 +1,893 @@
+					grb::config::ColIndexType
+				>::type
+			> parser_precond( data_in.precond_filename, data_in.direct );
+			if( parser_precond.m() != parser_precond.n() ) {
+				std::cerr << " matrix in " << data_in.precond_filename << " file, is not rectangular!";
+				rc = grb::ILLEGAL;
+			} else if( parser_precond.n() != n ) {
+				std::cerr << " Preconditioner P("<< parser_precond.n() << ") mast have same dimensions as matrix A(" << n << ") !\n";
+				rc = grb::ILLEGAL;
+			}
+			rc = rc ? rc : buildMatrixUnique(
+				P,
+				parser_precond.begin( SEQUENTIAL ),
+				parser_precond.end( SEQUENTIAL ),
+				SEQUENTIAL
+			);
+#ifdef DEBUG
+			if( rc == grb::SUCCESS ) {
+				std::cout << "Matrix P built from " << data_in.precond_filename << "file successfully\n";
+			}
+#endif
+		}
+
+	}
+
+	out.time_io += timer.time();
+	timer.reset();
+
+	// read RHS vector
+	if( data_in.rhs ) {
+#ifdef DEBUG
+		std::cout << "Reading RHS vector from " << data_in.rhs_filename << " file \n";
+#endif
+		std::vector< ScalarType > buffer( n, zero );
+		std::ifstream inFile( data_in.rhs_filename );
+		if( inFile.is_open() ) {
+			for(size_t i = 0; i < n; ++i) {
+				if( !( inFile >> buffer[ i ] ) ){
+					std::cerr << "Error reading from: " << data_in.rhs_filename << "\n";
+					rc = grb::ILLEGAL;
+					break;
+				};
+			}
+			inFile.close(); // cloose input file
+
+			rc = rc ? rc : grb::buildVector( b, buffer.begin(), buffer.end(), SEQUENTIAL );
+			if( rc != SUCCESS ) {
+				std::cout << "RHS vector: buildVector failed!\n ";
+			}
+		}
+		out.time_io += timer.time();
+		timer.reset();
+
+	} else {
+		grb::set( x, one );
+		grb::set( b, zero );
+		rc = rc ? rc : grb::mxv( b, A, x, ring );
+		grb::set( x, zero );
+		out.time_preamble += timer.time();
+		timer.reset();
+	}
+
+	// get RHS vector norm
+	ScalarType bnorm = zero;
+	rc = grb::set( temp, b );
+	if( grb::utils::is_complex< ScalarType >::value ) {
+		Vector< ScalarType > temp2( n );
+		rc = grb::set( temp2, zero );
+		rc = rc ? rc : grb::eWiseLambda(
+			[ &, temp ] ( const size_t i ) {
+				temp2[ i ] = grb::utils::is_complex< ScalarType >::conjugate( temp [ i ] );
+			}, temp
+		);
+		rc = rc ? rc : grb::dot( bnorm, temp, temp2, ring );
+	} else {
+		rc = rc ? rc : grb::dot( bnorm, temp, temp, ring );
+	}
+	bnorm =std::sqrt( bnorm );
+
+#ifdef DEBUG
+	std::cout << "RHS norm = " << std::abs( bnorm ) << " \n";
+
+	out.pinnedVector = PinnedVector< ScalarType >( b, SEQUENTIAL );
+	std::cout << "RHS vector = ";
+	for( size_t k = 0; k < 10; ++k ) {
+		const ScalarType &nonzeroValue = out.pinnedVector.getNonzeroValue( k );
+		std::cout << nonzeroValue << " ";
+	}
+	std::cout << " ...  ";
+	for( size_t k = n - 10; k < n; ++k ) {
+		const ScalarType &nonzeroValue = out.pinnedVector.getNonzeroValue( k );
+		std::cout << nonzeroValue << " ";
+	}
+	std::cout << "\n";
+#endif
+
+
+	out.time_preamble += timer.time();
+	timer.reset();
+
+	// todo: this loop shoud be moved into gmres() after
+	// hessolve is used from ALP
+
```

Comment:
I don't understand this comment -- please clarify

Reply:
Normally GMRES algorithm does both the Arnoldi and  the GMRES iterations internally, here we keep the GMRES iterations outside because we do not want to use the temporary hessolve within 'algorithm'. In principle we can move both things to the _.hpp_ file. What do you prefer @anyzelman ?

Reply:
Oh I see what you mean. I think the expectation of someone running `algorithms/gmres.hpp` that it happens end-to-end, so that would go towards the latter option. Then later (and ideally) we can call "ALP/Dense" at this point right?